ledgers/ccf/transaction_processor/pdo_tp.cpp

```diff
@@ -54,26 +55,58 @@ namespace ccfapp
         };
 
         //======================================================================================================
-        // register enclave expected measurements (member method)
-        auto set_contract_enclave_attestatation_verification_policy = [this](auto& ctx, const nlohmann::json& params) {
+        // register contract enclave attestation check flag (member method)
+        auto set_contract_enclave_attestation_check_flag = [this](auto& ctx, const nlohmann::json& params) {
 
-            const auto in = params.get<RegisterContractEnclaveAttestationVerificationPolicy::In>();
+            const auto in = params.get<RegisterContractEnclaveAttestionCheckFlag::In>();
 
-            // get the current measurements view
-            auto attestation_policy_view = ctx.tx.rw(attestation_policy_table);
+            // get the current view of contract_enclave_check_attestation_flag
+            auto check_attestation_flag_view = ctx.tx.rw(contract_enclave_check_attestation_flag);
 
             // collect the data to be stored
-            ContractEnclaveAttestationVerificationPolicy attestation_policy;
+            ContractEnclaveAttestionCheckFlag check_attestation_flag;
+            check_attestation_flag.check_attestation = in.check_attestation;
+
+            //store the data
+            check_attestation_flag_view->put(PDO_ENCLAVE_CHECK_ATTESTATION_FLAG, check_attestation_flag);
```

Comment:
This is where I think the a check is missing -- i.e., if the flag is set, it should not be modifiable later (you only modify the policy).

I understand that in the end this could be considered part of the policy of the consortium, who ultimately gets to decide what to check and how.
However, the flag still depends on the SIM/HW mode flag, which is something we use exclusively for testing. In production, the expectation is that there is always a policy, and it's up to the consortium to decide whether it is permissive or restrictive.

Reply:
the flag is part of the policy. The current policy is "consortium can FLIP the flag if they think they need to". The policy that you are suggesting, and what is implemented are both reasonable and valid EXAMPLE policies, i don't see any reason to pick one over the other for our purposes. In a real deployment, these have to be discussed, and the policy appropriate for the deployment needs to be implemented.

Reply:
> the flag is part of the policy. The current policy is "consortium can FLIP the flag if they think they need to". The policy that you are suggesting, and what is implemented are both reasonable and valid EXAMPLE policies, i don't see any reason to pick one over the other for our purposes.

Under a suitable level of abstraction, I agree.

> In a real deployment, these have to be discussed, and the policy appropriate for the deployment needs to be implemented.

This is the problem. In a real deployment, using this implementation, now there are 2 APIs, and the "policy" is actually two separate policies, which are treated as a composition within the PDO TP.

Reply:
but here we are debating the meaning of the two APIs. the fact is the deployment needs to deal with two APIs. there may be asks to tweak the meanings of these existing APIs. We moved from one API to two, since we feel this is more reasonable, and gives better control to the entities that deploy the system. Whether the abstractions of these APIs make sense in a particular deployment scenario is IMHO out of scope here.

Reply:
my suggestion... the consortium cannot change the type of ledger (its either simulation or its attested). however, the consortium can set/change the properties that constitute an acceptable enclave in attested mode. 

at a minimum, if you allow the policy to be changed then you MUST provide an api to check that policy. and even that is probably insufficient... you'll really need to allow transactions to be submitted with an expectation for the value of the policy (e.g. check the policy, then submit but the policy changed between check and submit)

Reply:
I think I have a proposal that meets our requirements as I am interpreting them.

1. keep the 2 APIs as they are; we may add later (or now) something like `get_policy`, update_policy`
2. consolidate the data model for the policy, namely switch from multiple KV pairs (one for expected measurements and one for the flag) to 1 KV pair stored in CCF. This KV pair will be something like `<PDO_ENCLAVE_ATTESTATION_POLICY, {check_flag: .., expected measurements: .., version: ..}. This is _the_ policy.
3. the set flag API owns the flag and implements a "set-once" semantics: `if no policy stored; create policy version 1; set flag; store it; else fail`. Essentially, once the ledger starts, the flag is set and will not be changed.
4. the set measurements API implements a conditional update of the policy (only, not the flag). It receives as input `new_measurements = {expected measurements: .., version: ..}`, it gets the `current_policy` from CCF, then `if new_policy.version == current_policy.version+1; then current_policy.measurements/version = new_policy.measurements/version; else fail "unexpected policy version"`.

Reply:
per discussion, the policy will be changed to "flag can be set only once". Thanks!